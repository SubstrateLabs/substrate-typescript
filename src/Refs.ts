import { v4 as uuidv4 } from "uuid";

type NodeLike = {
  id: string;
};

export type Prop = { t: "Key"; value: string } | { t: "Ref"; value: Ref };

export type Ref = {
  t: "Ref";
  id: string;
  node: NodeLike;
  props: Prop[];
};
export const isRef = (maybeRef: any): boolean => {
  return maybeRef.t === "Ref";
};

export type RefTable = Record<Ref["id"], Ref>;

export type RefFactory = {
  makeProxiedRef: (node: NodeLike, props?: any[]) => Ref;
  refs: RefTable;
  isRef: (maybeRef: Object) => boolean;
  getTarget: (ref: Ref) => Ref;
};

// This is used internally to allow Ref objects to be used as props via a special string. We might want to make this configurable, but also may be OK with a well-known obscure value.
const ID_PREFIX = "$$ID:";

// This is a "virtual" property that exists on Proxied Refs and used to get the unproxied Ref.
const TARGET_PROP = "$target";

export const makeFactory = (refs: RefTable = {}): RefFactory => {
  const id = () => `${ID_PREFIX}${uuidv4()}`;

  const makeProxiedRef = (node: NodeLike, props: any[] = []): any => {
    const ref: Ref = {
      t: "Ref",
      id: id(),
      node,
      props,
    };
    refs[ref.id] = ref;

    return new Proxy(ref, {
      has(target: Ref, key: any) {
        if (key === TARGET_PROP) return true; // NOTE: minor hack here (see getTarget)
        return key in target;
      },
      get(target: Ref, prop: any, _receiver: any) {
        if (prop === TARGET_PROP) return target; // NOTE: minor hack here (see getTarget)

        /**
         * When a "ref" is being used as a parameter to another ref such as in the case of:
         *
         *    let ref1 = a.b.c;
         *    let ref2 = d.e[ref1];
         *
         * When `e[ref1]` is called, internally JS will send `toPrimitive` to `ref1` in order
         * to pass it as a paramter to `e[]` - because proxy objects cannot be transformed
         * into primitive types, we're specifying here to use a function that resolves to the
         * internal `proxyId`. When `e[]` recieves a property access of a string that starts
         * with the special `$proxyId:` part it will use the actual `ref1` value using a
         * private lookup table.
         */
        if (prop === Symbol.toPrimitive) {
          return () => target.id;
        }

        if (typeof prop !== "string") throw new Error("is this possible?");

        const nextProp = prop.startsWith(ID_PREFIX)
          ? { t: "Ref", value: refs[prop] }
          : { t: "Key", value: prop };
        const props = [...ref.props, nextProp];

        return makeProxiedRef(target.node, props);
      },
    });
  };

  const isRef = (maybeRef: Object): boolean => {
    if (TARGET_PROP in maybeRef) return true;
    if ("t" in maybeRef) return maybeRef.t === "Ref";
    return false;
  };

  const getTarget = (maybeProxied: Ref): Ref => {
    if (TARGET_PROP in maybeProxied) {
      return maybeProxied.$target as Ref;
    } else {
      return maybeProxied;
    }
  };

  return { makeProxiedRef, getTarget, isRef, refs };
};
