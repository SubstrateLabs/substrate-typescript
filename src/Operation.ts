import { Ref } from "./Refs";

type NonEmptyArray<T> = T[] & { 0: T };

export type HasId = { id: string };
export type Edge<T extends HasId> = [T["id"], T["id"]];
type OpId = {
  __$$SB_GRAPH_OP_ID$$__: string;
};

export type OpSet = {
  id: string;
  class: string;
  op_stack: Operation[];
  op_graph_edges: Edge<OpSet>[];
  node_graph_edges: Edge<HasId>[];
  origin_node_id: HasId["id"];
};

export type Operation =
  | GetOp<GetItemArgs>
  | GetOp<GetAttrArgs>
  | StringConcatOp<ConcatArgs>;
type GetItemArgs = { accessor: "item"; key: null; op_id: OpSet["id"] };
type GetAttrArgs = { accessor: "attr"; key: string; op_id: null };
type GetArgs = GetItemArgs | GetAttrArgs;
type GetOp<T extends GetArgs> = {
  type: "get";
  args: T;
};
type ConcatDirection = "left" | "right";
type ConcatArgs = {
  op_id: OpSet["id"] | null;
  val: string | null;
  direction: ConcatDirection | null;
};
type StringConcatOp<T extends ConcatArgs> = {
  type: "string-concat";
  args: T;
};

const getItemOp = (opId: OpSet["id"]): GetOp<GetItemArgs> => ({
  type: "get",
  args: {
    accessor: "item",
    key: null,
    op_id: opId,
  },
});

const getAttrOp = (key: string): GetOp<GetAttrArgs> => ({
  type: "get",
  args: {
    accessor: "attr",
    key: key,
    op_id: null,
  },
});

// a type of function that generates ids
type NewId = () => string;

// Creates a list of Ops from the given Ref
export const refOps = (ref: Ref, id: NewId): NonEmptyArray<OpSet> => {
  let op: OpSet = {
    id: id(),
    class: "Node",
    op_stack: [] as Operation[],
    op_graph_edges: [] as Edge<OpSet>[],
    node_graph_edges: [] as Edge<HasId>[],
    origin_node_id: ref.node.id,
  };
  let ops: NonEmptyArray<OpSet> = [op];

  ref.props.forEach((prop) => {
    if (prop.t === "Ref") {
      const [refOp, ...restRefOps] = refOps(prop.value, () => id());
      ops = [refOp, ...restRefOps].concat(ops) as NonEmptyArray<OpSet>;

      op.op_graph_edges.push([refOp.id, op.id]);
      op.node_graph_edges.push([prop.value.node.id, ref.node.id]);

      op.op_stack.push(getItemOp(refOp.id));
    } else if (prop.t === "Key") {
      op.op_stack.push(getAttrOp(prop.value));
    }
  });

  return ops;
};

export const opId = (id: OpSet["id"]): OpId => ({
  __$$SB_GRAPH_OP_ID$$__: id,
});

export const replaceRefsWithOps = (args: any, refFactory: any, newId: any) => {
  let collectedOps: OpSet[] = [];

  const traverse = (obj: any): any => {
    if (Array.isArray(obj)) {
      return obj.map((item) => traverse(item));
    }

    if (typeof obj === "object") {
      if (refFactory.isRef(obj)) {
        const ref = refFactory.getTarget(obj);
        const ops = refOps(ref, newId);
        collectedOps = collectedOps.concat(ops);
        const entryOp = ops.at(-1);
        return opId(entryOp!.id);
      } else {
        return Object.keys(obj).reduce((a: any, k: any) => {
          a[k] = traverse(obj[k]);
          return a;
        }, {});
      }
    }

    return obj;
  };

  return {
    args: traverse(args),
    ops: collectedOps,
  };
};
