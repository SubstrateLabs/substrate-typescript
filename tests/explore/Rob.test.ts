import { expect, describe, test } from "vitest";
import * as Refs from "substrate/Refs";
import * as Operation from "substrate/Operation";
import { DiGraph } from "substrate/DiGraph";

const refFactory = Refs.makeFactory();

class FooNode {
  id: string;
  args: any;

  constructor(id: string, args: any = {}) {
    this.id = id;
    this.args = args;
  }

  get ref() {
    return refFactory.makeProxiedRef(this) as any;
  }

  toJSON() {
    return {
      class: "FooNode",
      id: this.id,
      args: this.args,
    };
  }
}

type NodeLike = { id: string; args: Object };

// simplified graph for demo
class Graph {
  newId: any;
  graph: DiGraph;

  constructor(newId: any, DAG: DiGraph = new DiGraph()) {
    this.newId = newId;
    this.graph = DAG;
  }

  add(node: NodeLike): Graph {
    this.graph.addNode([node.id, node]);

    const { ops } = Operation.replaceRefsWithOps(
      node.args,
      refFactory,
      this.newId
    );
    ops.forEach((op) => {
      this.graph.addEdge([op.origin_node, node.id, {}]);
    });

    return this;
  }

  get nodes() {
    return this.graph.nodes.map(([_nodeId, node]) => node);
  }

  get edges() {
    return this.graph.edges;
  }

  toJSON() {
    return this.nodes.reduce(
      (acc, node) => {
        const { args, ops } = Operation.replaceRefsWithOps(
          node.args,
          refFactory,
          this.newId
        );

        return {
          nodes: [...acc.nodes, { ...node.toJSON(), args }],
          ops: [...acc.ops, ...ops],
        };
      },
      { nodes: [], ops: [] }
    );
  }
}

// Rob's Example
const a = new FooNode("a");
const b = new FooNode("b", {
  foo: "b",
  bar: a.ref.foo,
  nested: [{ nest_id: "first" }, { nest_id: "second" }],
});
const c = new FooNode("c", {
  foo: "c",
  bar: b.ref.nested[a.ref.foo].nest_id,
  nested: { baz: b.ref.bar },
});

const idGenerator = (start: number = 1) => {
  let n = start;
  return () => {
    const id = n.toString();
    n = n + 1;
    return id;
  };
};

describe("Rob's Example", () => {
  test("a.ref.foo", () => {
    let result = Operation.refOps(
      refFactory.getTarget(b.args.bar),
      idGenerator()
    );
    expect(result).toEqual([
      {
        id: "1",
        op_stack: [
          { type: "get", args: { key: "foo", accessor: "attr", op_id: null } },
        ],
        op_graph_edges: [],
        node_graph_edges: [],
        origin_node: "a",
      },
    ]);
  });

  test("b.ref.nested[a.ref.foo].nest_id", () => {
    let result = Operation.refOps(
      refFactory.getTarget(c.args.bar),
      idGenerator()
    );
    expect(result).toEqual([
      {
        id: "2",
        op_stack: [
          {
            type: "get",
            args: { key: "foo", accessor: "attr", op_id: null },
          },
        ],
        op_graph_edges: [],
        node_graph_edges: [],
        origin_node: "a",
      },
      {
        id: "1",
        op_stack: [
          {
            type: "get",
            args: { key: "nested", accessor: "attr", op_id: null },
          },
          {
            type: "get",
            args: { key: null, accessor: "item", op_id: "2" },
          },
          {
            type: "get",
            args: { key: "nest_id", accessor: "attr", op_id: null },
          },
        ],
        op_graph_edges: [["2", "1"]],
        node_graph_edges: [["a", "b"]],
        origin_node: "b",
      },
    ]);
  });

  test("b.ref.bar", () => {
    let result = Operation.refOps(
      refFactory.getTarget(c.args.nested.baz),
      idGenerator()
    );
    expect(result).toEqual([
      {
        id: "1",
        op_stack: [
          {
            type: "get",
            args: { key: "bar", accessor: "attr", op_id: null },
          },
        ],
        op_graph_edges: [],
        node_graph_edges: [],
        origin_node: "b",
      },
    ]);
  });

  test("replace refs: c.args", () => {
    // const c = new FooNode("c", {
    //   foo: "c",
    //   bar: b.ref.nested[a.ref.foo].nest_id,
    //   nested: { baz: b.ref.bar },
    // });
    let { args, ops } = Operation.replaceRefsWithOps(
      c.args,
      refFactory,
      idGenerator()
    );

    expect(args).toEqual({
      foo: "c",
      bar: { __$$SB_GRAPH_OP_ID$$__: "1" },
      nested: { baz: { __$$SB_GRAPH_OP_ID$$__: "3" } },
    });

    expect(ops).toEqual([
      {
        id: "2",
        op_stack: [
          { type: "get", args: { accessor: "attr", key: "foo", op_id: null } },
        ],
        op_graph_edges: [],
        node_graph_edges: [],
        origin_node: "a",
      },
      {
        id: "1",
        op_stack: [
          {
            type: "get",
            args: { accessor: "attr", key: "nested", op_id: null },
          },
          { type: "get", args: { accessor: "item", key: null, op_id: "2" } },
          {
            type: "get",
            args: { accessor: "attr", key: "nest_id", op_id: null },
          },
        ],
        op_graph_edges: [["2", "1"]],
        node_graph_edges: [["a", "b"]],
        origin_node: "b",
      },
      {
        id: "3",
        op_stack: [
          { type: "get", args: { accessor: "attr", key: "bar", op_id: null } },
        ],
        op_graph_edges: [],
        node_graph_edges: [],
        origin_node: "b",
      },
    ]);
  });

  test("graph serialization (include ops, op-id replacement)", () => {
    const g = new Graph(idGenerator()).add(a).add(b).add(c);
    const result = g.toJSON();
    // console.log(JSON.stringify(result))
    expect(result.nodes.length).toEqual(3);
    expect(result.ops.length).toEqual(4);
    expect(result).toMatchSnapshot();
  });

  test("ok", () => {
    expect("ok").toEqual("ok");
  });
});
