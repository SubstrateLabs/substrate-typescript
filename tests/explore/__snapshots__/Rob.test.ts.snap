// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rob's Example > graph serialization (include ops, op-id replacement) 1`] = `
{
  "nodes": [
    {
      "args": {},
      "class": "FooNode",
      "id": "a",
    },
    {
      "args": {
        "bar": {
          "__$$SB_GRAPH_OP_ID$$__": "5",
        },
        "foo": "b",
        "nested": [
          {
            "nest_id": "first",
          },
          {
            "nest_id": "second",
          },
        ],
      },
      "class": "FooNode",
      "id": "b",
    },
    {
      "args": {
        "bar": {
          "__$$SB_GRAPH_OP_ID$$__": "6",
        },
        "foo": "c",
        "nested": {
          "baz": {
            "__$$SB_GRAPH_OP_ID$$__": "8",
          },
        },
      },
      "class": "FooNode",
      "id": "c",
    },
  ],
  "ops": [
    {
      "class": "Node",
      "id": "5",
      "node_graph_edges": [],
      "op_graph_edges": [],
      "op_stack": [
        {
          "args": {
            "accessor": "attr",
            "key": "foo",
            "op_id": null,
          },
          "type": "get",
        },
      ],
      "origin_node_id": "a",
    },
    {
      "class": "Node",
      "id": "7",
      "node_graph_edges": [],
      "op_graph_edges": [],
      "op_stack": [
        {
          "args": {
            "accessor": "attr",
            "key": "foo",
            "op_id": null,
          },
          "type": "get",
        },
      ],
      "origin_node_id": "a",
    },
    {
      "class": "Node",
      "id": "6",
      "node_graph_edges": [
        [
          "a",
          "b",
        ],
      ],
      "op_graph_edges": [
        [
          "7",
          "6",
        ],
      ],
      "op_stack": [
        {
          "args": {
            "accessor": "attr",
            "key": "nested",
            "op_id": null,
          },
          "type": "get",
        },
        {
          "args": {
            "accessor": "item",
            "key": null,
            "op_id": "7",
          },
          "type": "get",
        },
        {
          "args": {
            "accessor": "attr",
            "key": "nest_id",
            "op_id": null,
          },
          "type": "get",
        },
      ],
      "origin_node_id": "b",
    },
    {
      "class": "Node",
      "id": "8",
      "node_graph_edges": [],
      "op_graph_edges": [],
      "op_stack": [
        {
          "args": {
            "accessor": "attr",
            "key": "bar",
            "op_id": null,
          },
          "type": "get",
        },
      ],
      "origin_node_id": "b",
    },
  ],
}
`;
